File: server/src/lib/schemas.ts
import { z } from "zod";

export const loginSchema = z.object({
  email: z
    .string()
    .min(1, "Email or roll number is required")
    .max(255, "Identifier too long"),
  password: z
    .string()
    .min(1, "Password is required")
    .max(128, "Password too long"),
});

export const createUserSchema = z.object({
  name: z.string().min(1, "Name is required").max(100),
  email: z.string().email("Invalid email").max(255),
  password: z
    .string()
    .min(8, "Password must be at least 8 characters")
    .max(128),
  role: z.enum(["admin", "tpo", "recruiter", "candidate"]),
});

export const createRecruiterSchema = z.object({
  name: z.string().min(1, "Name is required").max(100),
  email: z.string().email("Invalid email").max(255),
  password: z.string().min(8, "Password must be at least 8 characters").max(128),
});

export const createTpoSchema = z.object({
  name: z.string().min(1, "Name is required").max(100),
  email: z.string().email("Invalid email").max(255),
  password: z.string().min(8, "Password must be at least 8 characters").max(128),
  college_name: z.string().min(1, "College name is required").max(200),
  college_code: z.string().min(1, "College code is required").max(50),
  location: z.string().max(200).optional(),
});

export const createCandidateSchema = z.object({
  name: z.string().min(1, "Name is required").max(100),
  email: z.string().email("Invalid email").max(255),
  password: z.string().min(8, "Password must be at least 8 characters").max(128),
  roll_number: z.string().max(50).optional(),
  college_id: z.string().uuid().optional(),
});

export const createExamSchema = z.object({
  title: z.string().min(1, "Title is required").max(200),
  description: z.string().max(2000).optional(),
  duration: z.number().int().min(5, "Duration must be at least 5 minutes"),
  total_marks: z.number().int().positive("Total marks must be positive"),
  pass_marks: z.number().int().min(0, "Pass marks cannot be negative"),
  available_from: z.string().datetime().optional(),
  available_until: z.string().datetime().optional(),
}).refine((data) => data.pass_marks <= data.total_marks, {
  message: "Pass marks cannot be greater than total marks",
  path: ["pass_marks"],
});

export const paginationSchema = z.object({
  page: z.string().optional().transform((v) => Math.max(1, Number(v) || 1)),
  limit: z.string().optional().transform((v) => {
    const n = Number(v) || 50;
    return Math.min(50, Math.max(1, n));
  }),
});

export const createJobSchema = z.object({
  title: z.string().min(1, "Job title is required").max(200),
  company_name: z.string().min(1, "Company name is required").max(200),
  description: z.string().max(5000).optional(),
  location: z.string().max(200).optional(),
  salary_range: z.string().max(100).optional(),
  job_type: z.enum(["full_time", "internship", "contract"]).default("full_time"),
  college_id: z.string().uuid().optional(),
  exam_id: z.string().uuid().optional(),
});

export const aiGenerateSchema = z.object({
  topic: z.string().min(1, "Topic is required").max(100),
  difficulty: z.enum(["easy", "medium", "hard", "very_hard"]).default("medium"),
  count: z.number().int().min(1).max(50).default(10),
  type: z.enum(["mcq", "coding"]).default("mcq"),
});

export const interviewAnswerSchema = z.object({
  interview_id: z.string().uuid(),
  question_index: z.number().int().min(0),
  answer_text: z.string().max(5000).optional(),
  audio_url: z.string().url().optional(),
});

export const proctoringEventSchema = z.object({
  attempt_id: z.string().uuid(),
  event_type: z.enum(["face_missing", "multiple_faces", "tab_switch", "copy_paste", "suspicious_movement"]),
  severity: z.enum(["low", "medium", "high"]).default("medium"),
  details: z.string().max(1000).optional(),
});

// ─── NEW SCHEMAS FOR ROUTE VALIDATION ───

export const startExamSchema = z.object({
  exam_id: z.string().uuid("exam_id must be a valid UUID"),
});

export const scheduleInterviewSchema = z.object({
  scheduled_start: z.string().datetime().optional(),
  scheduled_end: z.string().datetime().optional(),
});

export const interviewAnswerBodySchema = z.object({
  question: z.string().min(1, "Question is required").max(2000),
  answer: z.string().max(10000).optional(),
  stage: z.number().int().min(1).max(20).optional(),
});

export const submitMcqSchema = z.object({
  attempt_id: z.string().uuid("attempt_id must be a valid UUID"),
  question_id: z.string().uuid("question_id must be a valid UUID"),
  selected_option: z.string().min(1, "selected_option is required").max(1),
});

export const submitCodeSchema = z.object({
  attempt_id: z.string().uuid("attempt_id must be a valid UUID"),
  coding_question_id: z.string().uuid("coding_question_id must be a valid UUID"),
  code: z.string().max(50000, "Code exceeds maximum length"),
  language: z.string().min(1, "Language is required").max(50),
});

export const submitExamSchema = z.object({
  attempt_id: z.string().uuid("attempt_id must be a valid UUID"),
});

export const updateCodeScoreSchema = z.object({
  attempt_id: z.string().uuid("attempt_id must be a valid UUID"),
  coding_question_id: z.string().uuid("coding_question_id must be a valid UUID"),
  score: z.number().int().min(0).max(100),
  code: z.string().max(50000).optional(),
  language: z.string().max(50).optional(),
});

export const updateProfileSchema = z.object({
  phone: z.string().max(20).optional(),
  skills: z.array(z.string().max(100)).max(100).optional(),
  domain_preference: z.string().max(100).optional(),
  github_url: z.string().url().max(500).optional().or(z.literal("")),
  linkedin_url: z.string().url().max(500).optional().or(z.literal("")),
  portfolio_url: z.string().url().max(500).optional().or(z.literal("")),
  bio: z.string().max(2000).optional(),
  photo_url: z.string().url().max(500).optional().or(z.literal("")),
  projects: z.array(z.object({}).passthrough()).max(50).optional(),
  semester_grades: z.array(z.object({}).passthrough()).max(20).optional(),
});

export const onboardingSchema = z.object({
  password: z.string().min(8, "Password must be at least 8 characters").max(128),
  phone: z.string().min(1, "Phone is required").max(20),
  skills: z.array(z.string().min(1).max(100)).min(1, "At least one skill is required").max(50),
  domain_preference: z.string().min(1, "Domain preference is required").max(100),
  marksheet_url: z.string().url().max(500).optional().or(z.literal("")),
  resume_url: z.string().url().max(500).optional().or(z.literal("")),
});

export const proctoringLogEventSchema = z.object({
  attempt_id: z.string().uuid("attempt_id must be a valid UUID"),
  exam_id: z.string().uuid("exam_id must be a valid UUID"),
  event_type: z.enum(["camera_check", "snapshot", "violation", "submission"]),
  violation_count: z.number().int().min(0).optional(),
  message: z.string().max(2000).optional(),
  snapshot_data: z.string().max(10000000).optional(), // base64 snapshots can be large
});

export const respondOfferSchema = z.object({
  response: z.enum(["accept", "decline", "negotiate"]),
  notes: z.string().max(5000).optional().or(z.literal("")),
});

export const linkBankMcqSchema = z.object({
  exam_id: z.string().uuid(),
  question_ids: z.array(z.string().uuid()),
});

export const linkBankCodingSchema = z.object({
  exam_id: z.string().uuid(),
  coding_question_ids: z.array(z.string().uuid()),
});

export const addQuestionsSchema = z.object({
  exam_id: z.string().uuid(),
  questions: z.array(
    z.object({
      question_text: z.string().min(1).max(10000),
      option_a: z.string().min(1).max(5000),
      option_b: z.string().min(1).max(5000),
      option_c: z.string().min(1).max(5000),
      option_d: z.string().min(1).max(5000),
      correct_option: z.enum(["a", "b", "c", "d", "A", "B", "C", "D"]),
      marks: z.number().int().min(1),
    })
  ),
});

export const addCodingQuestionsSchema = z.object({
  exam_id: z.string().uuid(),
  coding_questions: z.array(
    z.object({
      title: z.string().min(1).max(255),
      description: z.string().min(1).max(20000),
      difficulty: z.enum(["easy", "medium", "hard"]),
      starter_code: z.string().max(50000),
      test_cases: z.array(
        z.object({
          input: z.string().max(10000),
          expected_output: z.string().max(10000),
        })
      ),
      marks: z.number().int().min(1),
    })
  ),
});

export const assignExamSchema = z.object({
  exam_id: z.string().uuid(),
  candidate_ids: z.array(z.string().uuid()),
});

export const uploadStudentsSchema = z.object({
  rows: z.array(
    z.object({
      roll_number: z.string().min(1).max(50),
      name: z.string().min(1).max(100),
      branch: z.string().min(1).max(100),
      cgpa: z.number().min(0).max(10),
      graduation_year: z.number().int().min(1900).max(2100),
      email: z.string().email().max(100).optional().or(z.literal("")),
    })
  ),
});

export const scanMarksheetsSchema = z.object({
  files: z.array(
    z.object({
      name: z.string().min(1).max(255),
      mimeType: z.string().min(1).max(100),
      data: z.string(),
    })
  ),
});

export const studentVerificationSchema = z.object({
  documents_verified: z.boolean(),
});

export const verifyStudentBatchSchema = z.object({
  studentIds: z.array(z.string().uuid()),
  documents_verified: z.boolean(),
});

export const assignDriveExamSchema = z.object({
  exam_id: z.string().uuid(),
});

export const saveDriveAiConfigSchema = z.object({
  aiConfig: z.object({}).passthrough(),
});

export const testDriveAiConfigSchema = z.object({
  question: z.string().min(1).max(5000),
  answer: z.string().min(1).max(20000),
  aiConfig: z.object({}).passthrough(),
});

export const aiShortlistSchema = z.object({
  criteria: z.string().min(1).max(5000),
});

export const resumeParseSchema = z.object({
  resume_text: z.string().min(1).max(1000000),
  job_skills: z.array(z.string().max(100)).max(100).optional(),
});

export const generateMcqSchema = z.object({
  topic: z.string().min(1).max(500),
  difficulty: z.enum(["easy", "medium", "hard"]),
  count: z.number().int().min(1).max(50),
});

export const generateCodingSchema = z.object({
  topic: z.string().min(1).max(500),
  difficulty: z.enum(["easy", "medium", "hard"]),
  count: z.number().int().min(1).max(50).optional(),
});

export const improvementReportSchema = z.object({
  attempt_id: z.string().uuid(),
});

export const runCodeSchema = z.object({
  code: z.string().max(50000),
  language: z.string().min(1).max(50),
  stdin: z.string().max(10000).optional(),
});

export const submitCompilerSchema = z.object({
  code: z.string().max(50000),
  language: z.string().min(1).max(50),
  test_cases: z.array(
    z.object({
      input: z.string().max(10000),
      expected_output: z.string().max(10000),
    })
  ),
});

export const snapshotOverrideSchema = z.object({
  violation_severity: z.enum(["low", "medium", "high", "critical"]),
});

export const addBankQuestionsSchema = z.object({
  questions: z.array(
    z.object({
      question_text: z.string().min(1).max(10000),
      option_a: z.string().min(1).max(5000),
      option_b: z.string().min(1).max(5000),
      option_c: z.string().min(1).max(5000),
      option_d: z.string().min(1).max(5000),
      correct_option: z.enum(["a", "b", "c", "d", "A", "B", "C", "D"]),
      marks: z.number().int().min(1).optional(),
    })
  ),
});

export const addBankCodingSchema = z.object({
  question: z.object({
    title: z.string().min(1).max(255),
    description: z.string().min(1).max(20000),
    difficulty: z.enum(["easy", "medium", "hard"]).optional(),
    starter_code: z.string().max(50000).optional(),
    test_cases: z.array(
      z.object({
        input: z.string().max(10000),
        expected_output: z.string().max(10000),
      })
    ).optional(),
    marks: z.number().int().min(1).optional(),
  }),
});

File: server_py/app/schemas.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, EmailStr, Field, model_validator


class LoginSchema(BaseModel):
    email: str = Field(..., min_length=1, max_length=255)
    password: str = Field(..., min_length=1, max_length=128)


class CreateUserSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr = Field(..., max_length=255)
    password: str = Field(..., min_length=8, max_length=128)
    role: str = Field(...)  # enum: admin, tpo, recruiter, candidate


class CreateRecruiterSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr = Field(..., max_length=255)
    password: str = Field(..., min_length=8, max_length=128)


class CreateTpoSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr = Field(..., max_length=255)
    password: str = Field(..., min_length=8, max_length=128)
    college_name: str = Field(..., min_length=1, max_length=200)
    college_code: str = Field(..., min_length=1, max_length=50)
    location: Optional[str] = Field(None, max_length=200)


class CreateCandidateSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr = Field(..., max_length=255)
    password: str = Field(..., min_length=8, max_length=128)
    roll_number: Optional[str] = Field(None, max_length=50)
    college_id: Optional[str] = None


class CreateExamSchema(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    duration: int = Field(..., ge=5)
    total_marks: int = Field(..., gt=0)
    pass_marks: int = Field(..., ge=0)
    available_from: Optional[str] = None
    available_until: Optional[str] = None

    @model_validator(mode="after")
    def validate_pass_marks(self) -> 'CreateExamSchema':
        if self.pass_marks > self.total_marks:
            raise ValueError("Pass marks cannot be greater than total marks")
        return self


class PaginationSchema(BaseModel):
    page: Optional[str] = None
    limit: Optional[str] = None


class CreateJobSchema(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    company_name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=5000)
    location: Optional[str] = Field(None, max_length=200)
    salary_range: Optional[str] = Field(None, max_length=100)
    job_type: str = "full_time"  # enum: full_time, internship, contract
    college_id: Optional[str] = None
    exam_id: Optional[str] = None


class AiGenerateSchema(BaseModel):
    topic: str = Field(..., min_length=1, max_length=100)
    difficulty: str = "medium"  # enum: easy, medium, hard, very_hard
    count: int = 10
    type: str = "mcq"  # enum: mcq, coding


class InterviewAnswerSchema(BaseModel):
    interview_id: str
    question_index: int = Field(..., ge=0)
    answer_text: Optional[str] = Field(None, max_length=5000)
    audio_url: Optional[str] = None


class ProctoringEventSchema(BaseModel):
    attempt_id: str
    event_type: str  # enum: face_missing, multiple_faces, tab_switch, copy_paste, suspicious_movement
    severity: str = "medium"  # enum: low, medium, high
    details: Optional[str] = Field(None, max_length=1000)


class StartExamSchema(BaseModel):
    exam_id: str


class ScheduleInterviewSchema(BaseModel):
    scheduled_start: Optional[str] = None
    scheduled_end: Optional[str] = None


class InterviewAnswerBodySchema(BaseModel):
    question: str = Field(..., min_length=1, max_length=2000)
    answer: Optional[str] = Field(None, max_length=10000)
    stage: Optional[int] = Field(None, ge=1, le=20)


class SubmitMcqSchema(BaseModel):
    attempt_id: str
    question_id: str
    selected_option: str = Field(..., min_length=1, max_length=1)


class SubmitCodeSchema(BaseModel):
    attempt_id: str
    coding_question_id: str
    code: str = Field(..., max_length=50000)
    language: str = Field(..., min_length=1, max_length=50)


class SubmitExamSchema(BaseModel):
    attempt_id: str


class UpdateCodeScoreSchema(BaseModel):
    attempt_id: str
    coding_question_id: str
    score: int = Field(..., ge=0, le=100)
    code: Optional[str] = Field(None, max_length=50000)
    language: Optional[str] = Field(None, max_length=50)


class UpdateProfileSchema(BaseModel):
    phone: Optional[str] = Field(None, max_length=20)
    skills: Optional[List[str]] = None
    domain_preference: Optional[str] = Field(None, max_length=100)
    github_url: Optional[str] = Field(None, max_length=500)
    linkedin_url: Optional[str] = Field(None, max_length=500)
    portfolio_url: Optional[str] = Field(None, max_length=500)
    bio: Optional[str] = Field(None, max_length=2000)
    photo_url: Optional[str] = Field(None, max_length=500)
    projects: Optional[List[Dict[str, Any]]] = None
    semester_grades: Optional[List[Dict[str, Any]]] = None


class OnboardingSchema(BaseModel):
    password: str = Field(..., min_length=8, max_length=128)
    phone: str = Field(..., min_length=1, max_length=20)
    skills: List[str] = Field(...)
    domain_preference: str = Field(..., min_length=1, max_length=100)
    marksheet_url: Optional[str] = Field("", max_length=500)
    resume_url: Optional[str] = Field("", max_length=500)


class ProctoringLogEventSchema(BaseModel):
    attempt_id: str
    exam_id: str
    event_type: str  # enum: camera_check, snapshot, violation, submission
    violation_count: Optional[int] = Field(None, ge=0)
    message: Optional[str] = Field(None, max_length=2000)
    snapshot_data: Optional[str] = Field(None, max_length=10000000)


class RespondOfferSchema(BaseModel):
    response: str  # enum: accept, decline, negotiate
    notes: Optional[str] = Field("", max_length=5000)


class LinkBankMcqSchema(BaseModel):
    exam_id: str
    question_ids: List[str]


class LinkBankCodingSchema(BaseModel):
    exam_id: str
    coding_question_ids: List[str]


class QuestionItemSchema(BaseModel):
    question_text: str = Field(..., min_length=1, max_length=10000)
    option_a: str = Field(..., min_length=1, max_length=5000)
    option_b: str = Field(..., min_length=1, max_length=5000)
    option_c: str = Field(..., min_length=1, max_length=5000)
    option_d: str = Field(..., min_length=1, max_length=5000)
    correct_option: str = Field(...)  # enum: a, b, c, d, A, B, C, D
    marks: int = Field(..., ge=1)


class AddQuestionsSchema(BaseModel):
    exam_id: str
    questions: List[QuestionItemSchema]


class TestCaseItemSchema(BaseModel):
    input: str = Field(..., max_length=10000)
    expected_output: str = Field(..., max_length=10000)


class CodingQuestionItemSchema(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str = Field(..., min_length=1, max_length=20000)
    difficulty: str  # enum: easy, medium, hard
    starter_code: str = Field(..., max_length=50000)
    test_cases: List[TestCaseItemSchema]
    marks: int = Field(..., ge=1)


class AddCodingQuestionsSchema(BaseModel):
    exam_id: str
    coding_questions: List[CodingQuestionItemSchema]


class AssignExamSchema(BaseModel):
    exam_id: str
    candidate_ids: List[str]


class StudentRowSchema(BaseModel):
    roll_number: str = Field(..., min_length=1, max_length=50)
    name: str = Field(..., min_length=1, max_length=100)
    branch: str = Field(..., min_length=1, max_length=100)
    cgpa: float = Field(..., ge=0, le=10)
    graduation_year: int = Field(..., ge=1900, le=2100)
    email: Optional[str] = Field("", max_length=100)


class UploadStudentsSchema(BaseModel):
    rows: List[StudentRowSchema]


class FileItemSchema(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    mimeType: str = Field(..., min_length=1, max_length=100)
    data: str


class ScanMarksheetsSchema(BaseModel):
    files: List[FileItemSchema]


class StudentVerificationSchema(BaseModel):
    documents_verified: bool


class VerifyStudentBatchSchema(BaseModel):
    studentIds: List[str]
    documents_verified: bool


class AssignDriveExamSchema(BaseModel):
    exam_id: str


class SaveDriveAiConfigSchema(BaseModel):
    aiConfig: Dict[str, Any]


class TestDriveAiConfigSchema(BaseModel):
    question: str = Field(..., min_length=1, max_length=5000)
    answer: str = Field(..., min_length=1, max_length=20000)
    aiConfig: Dict[str, Any]


class AiShortlistSchema(BaseModel):
    criteria: str = Field(..., min_length=1, max_length=5000)


class ResumeParseSchema(BaseModel):
    resume_text: str = Field(..., min_length=1, max_length=1000000)
    job_skills: Optional[List[str]] = None


class GenerateMcqSchema(BaseModel):
    topic: str = Field(..., min_length=1, max_length=500)
    difficulty: str  # enum: easy, medium, hard
    count: int = Field(..., ge=1, le=50)


class GenerateCodingSchema(BaseModel):
    topic: str = Field(..., min_length=1, max_length=500)
    difficulty: str  # enum: easy, medium, hard
    count: Optional[int] = Field(None, ge=1, le=50)


class ImprovementReportSchema(BaseModel):
    attempt_id: str


class RunCodeSchema(BaseModel):
    code: str = Field(..., max_length=50000)
    language: str = Field(..., min_length=1, max_length=50)
    stdin: Optional[str] = Field("", max_length=10000)


class SubmitCompilerSchema(BaseModel):
    code: str = Field(..., max_length=50000)
    language: str = Field(..., min_length=1, max_length=50)
    test_cases: List[TestCaseItemSchema]


class SnapshotOverrideSchema(BaseModel):
    violation_severity: str  # enum: low, medium, high, critical


class BankQuestionItemSchema(BaseModel):
    question_text: str = Field(..., min_length=1, max_length=10000)
    option_a: str = Field(..., min_length=1, max_length=5000)
    option_b: str = Field(..., min_length=1, max_length=5000)
    option_c: str = Field(..., min_length=1, max_length=5000)
    option_d: str = Field(..., min_length=1, max_length=5000)
    correct_option: str = Field(...)  # enum: a, b, c, d, A, B, C, D
    marks: Optional[int] = Field(None, ge=1)


class AddBankQuestionsSchema(BaseModel):
    questions: List[BankQuestionItemSchema]


class BankCodingQuestionSchema(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str = Field(..., min_length=1, max_length=20000)
    difficulty: Optional[str] = None  # enum: easy, medium, hard
    starter_code: Optional[str] = Field(None, max_length=50000)
    test_cases: Optional[List[TestCaseItemSchema]] = None
    marks: Optional[int] = Field(None, ge=1)


class AddBankCodingSchema(BaseModel):
    question: BankCodingQuestionSchema
